### PR TITLE
Accept client_id parameter for Azure ManagedIdentity auth to support …

### LINKED ISSUE #5745: Accept client_id parameter for Azure ManagedIdentity auth to support multiple identities
## Proposal

For [`azure_sd_config`](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#azure_sd_config), Prometheus supports either `OAuth` or `ManagedIdentity` as authentication method. I would like to request having an option to pick which Identity shall be used to authenticate against Azure for listing VMs and scraping metrics, in case the VM has multiple Identities assigned. 

The API distinguishes them by their `client_id` ([source](https://docs.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/overview#how-a-user-assigned-managed-identity-works-with-an-azure-vm); emphasis mine):

> 5. Your code that's running on the VM can request a token from the Azure Instance Metadata Service identity endpoint, accessible only from within the VM: `http://169.254.169.254/metadata/identity/oauth2/token`
>
>   - The resource parameter specifies the service to which the token is sent. To authenticate to Azure Resource Manager, use `resource=https://management.azure.com/`.
>   - **The client ID parameter specifies the identity for which the token is requested. This value is required for disambiguation when more than one user-assigned identity is on a single VM.**
>   - The API version parameter specifies the Azure Instance Metadata Service version. Use `api-version=2018-02-01` or higher.

Therefore the already existing option `client_id` should not be ignored for `ManagedIdentity` auth, but sent along as query parameter to allow selecting a specific Identity and thereby support machines having more than one. This should be a very minor change.

**Use case. Why is this important?**

If you want to set up Prometheus on a VM with anything but exactly one Managed Identity assigned, you can currently not use that authentication method but have to fall back to `OAuth`, including full credentials in your scrape config. 

On VMs with multiple Managed Identities (e.g. with different permissions for different purposes) right now the discovery fails with:

```
level=error ts=2019-07-09T12:48:59.620Z caller=refresh.go:95 component="discovery manager scrape" discovery=azure msg="Unable to refresh target groups" err="could not get virtual machines: could not list virtual machines: azure.BearerAuthorizer#WithAuthorization: Failed to refresh the Token for request to https://management.azure.com/subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/providers/Microsoft.Compute/virtualMachines?api-version=2018-10-01: StatusCode=400 -- Original Error: adal: Refresh request failed. Status Code = '400'. Response body: {\"error\":\"invalid_request\",\"error_description\":\"Identity not found\"}"
```

This can also be reproduced with plain `curl` requesting the Azure Instance Metadata Service endpoint from within an Azure VM that has multiple Identities:

```bash
$  curl 'http://169.254.169.254/metadata/identity/oauth2/token?api-version=2018-02-01&resource=https%3A%2F%2Fmanagement.azure.net' -H Metadata:true 
{"error":"invalid_request","error_description":"Identity not found"}

$ curl 'http://169.254.169.254/metadata/identity/oauth2/token?api-version=2018-02-01&resource=https%3A%2F%2Fmanagement.azure.net&client_id=yyyyyyyy-yyyy-yyyy-yyyy-yyyyyyyyyyyy' -H Metadata:true 
{"access_token":"abcdefg......", ...
```

### PR DESCRIPTION
**Pull request by qsbao**
…multiple identities

Closes #5745 

Signed-off-by: qsbao <qingsheng.bqs@gmail.com>

### CODE REVIEW

File: docs/configuration/configuration.md [hunk 1]
```diff
@@ -294,7 +294,7 @@ See below for the configuration options for Azure discovery:
 subscription_id: <string>
 # Optional tenant ID. Only required with authentication_method OAuth.
 [ tenant_id: <string> ]
-# Optional client ID. Only required with authentication_method OAuth.
+# Optional client ID. Required with authentication_method OAuth, optional for authentication_method ManagedIdentity.
```
Comment (simonpasquier) [inline docs/configuration/configuration.md]: I would shorten it: `Optional client ID. Required with authentication_method OAuth.`

Comment (qsbao) [reply docs/configuration/configuration.md]: All right.

**Comment by simonpasquier** [comment]
@ByteCommander as the original reporter, maybe you can test this patch?

**Comment by ByteCommander** [comment]
Sorry for the delay, work is a bit busy right now and we decided to drop the use case for that scrape config temporarily. I'll try to get back to it soon though. Thank you for your responses.

**Comment by beorn7** [comment]
Any updates on this?

**Comment by roidelapluie** [comment]
Ping @simonpasquier is this in a mergeable state?

**Comment by simonpasquier** [comment]
The code looks good to me but I can't tell whether it's fixing the original issue or not.

**Comment by brian-brazil** [comment]
@ByteCommander Did you get a chance to test this out?

**Comment by ByteCommander** [comment]
Unfortunately I am no longer working on this project and have no ability to test the issue any more, sorry.

**Comment by gbavoz1** [comment]
I found this issue again, specially now with Pod-Identity enabled clusters that access multiple identities. Can this be merged or needs more work (I might be able to contribute)?

**Comment by arnitolog** [comment]
Hello, I can confirm that with this change in place and with option `client_id` in azure_sd configuration, Prometheus was able to select the propper Managed Identity and discovery works as expected. 
What is the next step to have this change in the next release? 
cc: @brian-brazil , @simonpasquier , @roidelapluie

**Comment by roidelapluie** [comment]
A rebase would be great to ensure that CI passes.

**Comment by dsiperek-vendavo** [comment]
Without this option you can't use Application Gateway Ingress controller as it creates another managed identity!

**Comment by qsbao** [comment]
Sorry I haven't been working on this for a long time, maybe I should just close this Pull Request.

**Pull request closed without merging**